### ISSUE TITLE
Weird render glitches with Kitty

### ISSUE THREAD

**Issue by benjamin051000** [1j01/textual-paint#4]
See how there is one row with some strange artifacts after starting up? Using latest release of Kitty. They go away when I move a window (e.g., About window) over the artifacts.

![image](https://github.com/1j01/textual-paint/assets/34549440/72bbd845-fb07-4c13-855e-b143c3ca88b1)

**Comment by georgewoodall82** [comment]
alacritty

**Comment by 1j01** [comment]
Hi, I ran into these sorts of issues on each of the platforms I tested, and solved it by [picking different emoji for different terminals](https://github.com/1j01/textual-paint/blob/ec63ff0e5598bda3710cc7ac456e0f4df602f599/src/textual_paint/paint.py#L245). I was also planning on adding a way to customize the symbols used to represent the tools, since I expect these issues will continue to crop up on different platforms, or depending on font customization.

The problem as I understand it is that character widths can't be easily measured in the terminal, and so Rich, the library underlying the Textual framework that Textual Paint uses, relies on a [pre-computed](https://github.com/Textualize/rich/blob/720800e6930d85ad027b1e9bd0cbb96b5e994ce3/tools/make_terminal_widths.py#L13) table of widths for characters, using the [wcwidth](https://pypi.org/project/wcwidth/) module (which itself is defined by fixed tables.)
Then when Textual renders the screen, any time it updates a row of the screen, it bundles together anything to the right of what it needs to update. This might not be such a problem if the toolbox was on the right, but since it's on the left (and has two columns of tools), the incorrect assumption about the character width affects everything to the right of it, creating a ragged effect.
When you move a window, each row is redrawn using absolute coordinates to jump to the left of the update region, so it fixes it temporarily.

Two options I see for fixing this, other than the workaround of using different emoji:
- If there were a way to mark a widget as an unknowable width, and then the Textual framework would always use absolute coordinate jumps to render anything to the right, this could fix it similar to how a dialog window does. This would be a very limited solution though, as it wouldn't work for Rich's text wrapping/centering, or `auto` width widgets.
- Since it is actually possible to measure the cell width of a character, at least on most terminals, using a Device Status Report control code to get the cursor position before and after writing the character to the terminal (or perhaps just after, if making sure to start at the beginning of a line), even though this is might not be practical in general for computing the cell width of characters, if I could do this just for the tool icons, and then _patch_ that table to be more accurate, that might solve it nicely. That said, over a remote SSH connection, it would theoretically require 16 _round trips_ at startup time. That's certainly _stupid_, although for an application meant for interactive drawing (i.e. wanting low latency anyways), perhaps not prohibitive.

Anyways, in the short term it should be easy to add an extra conditional to handle Kitty or Alacritty, assuming they're detectable with `TERM_PROGRAM`, and in the even shorter term, there's `--ascii-only-icons` :)

**Comment by 1j01** [comment]
Okay, so I went ahead and tried to implement the long term (or, at least, theoretically universal) solution of patching the `CHAR_WIDTHS` table, but, after implementing it, I went to test this in VS Code's integrated terminal, and it turns out that "🪣" is measured as 2 cells wide with the SGR strategy, same as `wcwidth` gives, whereas it seems to only advance by 1 cell visually. It's very buggy when typing 🪣 at the command line in VS Code, whereas it's sensible in Ubuntu's default Terminal app, so I think this is a problem with VS Code, and it may or may not be a similar story for other terminals, but this certainly doesn't work as a universal fix. I'll leave the code here for future reference.

```py
"""Workaround for rich.cells.cell_len measuring some characters incorrectly."""

import curses
from typing import cast
from rich._cell_widths import CELL_WIDTHS

CELL_WIDTHS = cast(list[tuple[int, int, int]], CELL_WIDTHS)  # type: ignore
"""A list of (start, end, width) tuples, where start and end form an inclusive range of codepoints having that width."""

def accurate_slow_cell_len(text: str) -> int:
    """Measure the width of a string in cells, manually, using DSR control codes.

    `rich.cells.cell_len` can be used in general, and is fast, using a table and caching,
    but some characters are measured incorrectly, causing misalignment of everything to the right of them,
    since the width varies by terminal/locale/font.

    This function uses a Device Status Report control code to get the cursor position after writing to the terminal,
    and measures the difference.
    
    It should be used sparingly, at startup, before `cell_len` is used to measure the characters in question,
    to avoid caching the wrong values.
    """
    curses.setupterm()

    stdscr = curses.initscr()

    try:
        # Draw the text at the top left of the screen
        stdscr.addstr(0, 0, text)

        # Get the ending cursor position
        _, end_col = stdscr.getyx()

        # The width is the difference between the starting and ending cursor positions,
        # and since the starting cursor position is always 0, it's just the ending cursor position.
        width = end_col

    finally:
        # End the curses session
        curses.endwin()

    return width

def patch_cell_widths_table(text: str) -> None:
    """Patch the cell length measurements table for tool icons, since it varies by terminal/locale/font.
    
    This should be called before the characters in question are ever measured with cell_len, to avoid caching the wrong values.
    """
    for char in text:
        codepoint = ord(char)
        tuple_for_char = (codepoint, codepoint, accurate_slow_cell_len(char))
        patched = False
        for i in range(len(CELL_WIDTHS)):
            start, end, width = CELL_WIDTHS[i]
            if codepoint >= start and codepoint <= end:
                # Break up the range into three ranges, and fix the width of the middle one targeting char.
                CELL_WIDTHS[i:i+1] = [
                    (start, codepoint - 1, width),
                    tuple_for_char,
                    (codepoint + 1, end, width),
                ]
                patched = True
                break
        if not patched:
            # Add a new range for char. This needs to be inserted in order,
            # since _get_codepoint_cell_size uses a binary search.
            for i in range(len(CELL_WIDTHS)):
                start, end, width = CELL_WIDTHS[i]
                if codepoint < start:
                    CELL_WIDTHS.insert(i, tuple_for_char)
                    break
            else:
                CELL_WIDTHS.append(tuple_for_char)


if __name__ == '__main__':
    from rich.cells import cell_len

    TEST_CHARS = 'aあ✏️🖌️🪣🫗👩🏻‍🦰'

    # Compare the two functions
    # THIS TEST IS MUTUALLY EXCLUSIVE WITH THE PATCHING TEST BELOW
    # SINCE `cell_len` USES A CACHE!
    # for char in TEST_CHARS:
    #     print(f'{char} ({char.encode("unicode_escape")}) cell_len: {cell_len(char)} accurate_slow_cell_len: {accurate_slow_cell_len(char)}')

    # print(f'{TEST_CHARS} ({TEST_CHARS.encode("unicode_escape")}) cell_len: {cell_len(TEST_CHARS)} accurate_slow_cell_len: {accurate_slow_cell_len(TEST_CHARS)}')
    
    # Test that patching the table brings the two functions in line
    for char in TEST_CHARS:
        patch_cell_widths_table(char)
        expected = accurate_slow_cell_len(char)
        actual = cell_len(char)
        assert actual == expected, f'Expected {expected} for {char} ({char.encode("unicode_escape")}), got {actual}'
    
    expected = accurate_slow_cell_len(TEST_CHARS)
    actual = cell_len(TEST_CHARS)
    assert actual == expected, f'Expected {expected} for {TEST_CHARS} ({TEST_CHARS.encode("unicode_escape")}), got {actual}'

```
```py
def fix_tool_icon_cell_widths():
    """Patch the cell length measurements table for tool icons, since it varies by terminal/locale/font.
    
    This should be called before the characters in question are ever measured with cell_len, to avoid caching the wrong values.
    """
    if ascii_only_icons:
        return
    for tool in Tool:
        patch_cell_widths_table(tool.get_icon())

fix_tool_icon_cell_widths()
```

**Comment by 1j01** [comment]
## Kitty

I added icon swaps for Kitty. Nearly all the alternative emoji I would want to use for Pencil and Brush cause problems in Kitty as well (✏️, 🖍️, 🖊️, 🖋️, ✍️, ✒️, 🖎, 🖆, ✎,🪈, 🖌️, 🧹, 🪮), but I managed to make a new Brush icon out of two characters ( ▬▤) that I think looks pretty good, in color.
The closest working Pencil emoji is 📝, which may be confused with the Text tool, but at least it contains a pencil.

I also swapped out the Text tool and Curve tool icons for ones that look better in Kitty (for me at least.)

These improvements are in the `main` branch, but not published to PyPI yet.

![Screenshot of Kitty Terminal running Textual Paint with new tool icons](https://github.com/1j01/textual-paint/assets/3630663/df5554a5-d886-4411-a084-ba0dbd53001f)

## Alacritty

I installed Alacritty from Ubuntu's Software app since the `cargo` install failed. I see a lot more monochrome icons than on other terminals, but no alignment issues.

@georgewoodall82 Please open a new issue with details about Alacritty, and I'll take another look.

## Lowest Common Denominator

I also improved the lowest common denominator (Xterm, etc.) by adding color to the ASCII-only tool icons, and extending `--ascii-only-icons` to an `--ascii-only` option that affects the whole UI.

![Screenshot of Xterm running Textual Paint in ASCII only mode](https://github.com/1j01/textual-paint/assets/3630663/665b92c9-6558-466c-a6ad-603d86084ea5)


## Universal Solution

**tl;dr:** some hope but no concrete results

I thought of two more (related) avenues for possibly fixing this for all terminals: 1. position the tools absolutely, to hopefully trigger it render like it does when a tooltip is visible, 2. create an absolutely positioned widget, similar to a tooltip but always on screen (and ideally invisible), to hopefully trigger everything to render like when a tooltip is visible.

I tried option 2 since it seemed more likely to work, being closer to a tooltip. I copied the code for the `Tooltip` widget, renamed it `AlignmentFixer`, and added it to the DOM. Depending on where I put it in the DOM, however, it either *does nothing*, or *makes the tooltip not even fix the rendering temporarily.* So it does the opposite of what I wanted to achieve, essentially.

BEFORE, WITHOUT `AlignmentFixer`:
[Screencast without `AlignmentFixer`, showing tooltip temporarily fixing the misalignment](https://github.com/1j01/textual-paint/assets/3630663/fa762cec-0c75-4e60-b366-08944e9dd34e)

AFTER, with `AlignmentFixer`™:
[Screencast with `AlignmentFixer`. Now tooltips do nothing.](https://github.com/1j01/textual-paint/assets/3630663/b6ffc4a5-9ed3-4f15-be2b-ed2baf3eac64)

This is amusing and somewhat disheartening, but it's still a potentially viable avenue. I'd just need to replicate the `Tooltip` *even closer than copying the code for it*, somehow, eliminate all differences until I have a baseline to work from that has the same positive effect of the tooltip. It may need to be on the very top layer and placed exactly first (or last) in the DOM, and it may not work when a `Tooltip` is taking that spot. That could be a flaw in the solution, since I do want tooltips, but it may be solvable.

Maybe just setting a `tooltip` on the root widget, to some invisible text could work. I'm not sure if any text would be invisible and not taken to mean "don't show a tooltip"... and, well, I don't want a delay, and the tooltips have a delay... but maybe if it's set on the root node, the delay will only happen once since you're always hovering over it? But...

I don't actually know yet whether it's the existence of the tooltip or the tooltip transitioning from hidden to visible that's causing it to render nicely.
It's also worth investigating whether this workaround limits the rendering performance.

**Comment by 1j01** [comment]
Just published [0.2.0](https://pypi.org/project/textual-paint/0.2.0/) with the improved Kitty support and new `--ascii-only` mode.

(I also made some better ANSI art of the Kitty logo:)
<img alt="kitty terminal logo ANSI art" width="200" src="https://github.com/1j01/textual-paint/assets/3630663/bdb90ecd-9179-4c74-af01-6529cf28e3a8">
